### DISCUSSION TITLE
TypeError with ModelStream Object in Python

### DISCUSSION THREAD

**Discussion by snowyday** [guidance-ai/guidance#808] [Q&A]
I'm encountering a TypeError when attempting to process a guidance model stream with the OpenAI API in Python. Here's the relevant code:

```python
lm =  models.OpenAI("gpt-3.5-turbo-instruct", echo=False) + "Hi!"

step_output = ""
for event in lm.stream() + gen(name="response", temperature=0.1):
    post_text = event.get("response", "")
    for word in post_text[len(step_output) :]:
        step_output += word
        
 lm["response"]
 >> TypeError: 'ModelStream' object is not subscriptable
```

When executing this code, I get the following error: TypeError: 'ModelStream' object is not subscriptable.
Could someone clarify why this error occurs and how to properly access the 'response' in a ModelStream object?

**Comment by obragaa** [comment]
Hii snowyday,

The error `TypeError: 'ModelStream' object is not subscriptable` occurs because you are attempting to use the `ModelStream` object as if it were a dictionary, which it is not designed to be. In your code, the expression `lm["response"]` is what's causing this issue because `lm` is a `ModelStream` instance, which doesn't support dictionary-like access.

Here’s how you can correctly handle the streaming and access the response from an OpenAI model in Python:

```python
from openai import OpenAI

# Create an instance of the model
lm = OpenAI("gpt-3.5-turbo-instruct", echo=False)

# Start the stream by directly adding the initial message to the stream call
stream = lm.stream(gen(name="response", temperature=0.1, prompt="Hi!"))

# Variable to accumulate the output
step_output = ""

# Iterate over the events from the stream
for event in stream:
    # Access the response if it's available in the event
    if 'response' in event:
        post_text = event['response']
        for word in post_text[len(step_output):]:
            step_output += word

# step_output now contains the entire accumulated response from the model
print(step_output)
```

In this revised code, the streaming process is initiated with a direct call to `stream()` passing the necessary parameters including the initial prompt. The loop then iterates over the events returned by the stream, and for each event that contains a 'response', the text is processed and accumulated in `step_output`.

Make sure you are using the correct libraries and methods as per the OpenAI API documentation to avoid other potential issues.

Hope this helps!

@obragaa

**Reply by snowyday** [reply]
I now understand the cause of the error and how the Model Stream works. I've already updated my code accordingly.

Thank you!
